linera-chain/src/chain.rs

```diff
@@ -572,26 +574,58 @@ where
-        let entry = InboxEntry::new(origin.clone(), &bundle);
-        let skippable = bundle.is_skippable();
-        let newly_added = inbox
-            .add_bundle(bundle)
-            .await
-            .map_err(|error| match error {
-                InboxError::ViewError(error) => ChainError::ViewError(error),
-                error => ChainError::InternalError(format!(
-                    "while processing messages in certified block: {error}"
-                )),
-            })?;
-        if newly_added && !skippable {
-            let seen = local_time;
-            self.unskippable
-                .push_back(TimestampedInboxEntry { entry, seen });
+        if bundle.messages.iter().all(|posted_message| {
+            matches!(
+                posted_message.message,
+                Message::System(
+                    SystemMessage::Subscribe { .. } | SystemMessage::Unsubscribe { .. }
+                )
+            )
+        }) {
+            let names_and_ids = bundle
+                .messages
+                .iter()
+                .filter_map(|posted_message| posted_message.message.unsubscribe())
+                .map(|(id, subscription)| (subscription.name.clone(), *id))
+                .collect();
+            self.unsubscribe(names_and_ids, GenericApplicationId::System)
+                .await?;
+
+            let names_and_ids = bundle
+                .messages
+                .iter()
+                .filter_map(|posted_message| posted_message.message.subscribe())
+                .map(|(id, subscription)| (subscription.name.clone(), *id))
+                .collect();
+            self.subscribe(names_and_ids, GenericApplicationId::System, true)
+                .await?;
+        } else {
+            // Process the inbox bundle and update the inbox state.
+            let mut inbox = self.inboxes.try_load_entry_mut(origin).await?;
+            let entry = InboxEntry::new(origin.clone(), &bundle);
+            let skippable = bundle.is_skippable();
+            let newly_added = inbox
+                .add_bundle(bundle)
+                .await
+                .map_err(|error| match error {
+                    InboxError::ViewError(error) => ChainError::ViewError(error),
+                    error => ChainError::InternalError(format!(
+                        "while processing messages in certified block: {error}"
+                    )),
+                })?;
+            if newly_added && !skippable {
+                let seen = local_time;
+                self.unskippable
+                    .push_back(TimestampedInboxEntry { entry, seen });
+            }
         }
         // Remember the certificate for future validator/client synchronizations.
-        self.received_log.push(chain_and_height);
+        let log_count = self.received_log.count();
+        if log_count == 0
+            || self.received_log.get(log_count - 1).await?.as_ref() != Some(&chain_and_height)
```

Comment:
If we receive multiple bundles from the same sending block, they will always be added as consecutive entries in `received_log`, so this would be effective.

I can move this logic to the call site instead. (Or remove it entirely, if you prefer. Maybe multiple bundles from the same sender is actually not a common use case.)

Reply:
No because you could receive a block from the past after you just subscribed to a channel

Reply:
Right, not always.
It only helps in the case where a single block is sending multiple messages to us that are all either direct or via channels we are already subscribed to.